Document the concept of “experimental feature” (#5930)
NixOS/nix: doc/manual/src/contributing/experimental-flags.md

[Since Nix 2.4](../release-notes/rl-2.4.md), some features of Nix can be flagged as experimental.
Experimental features can be changed or removed at any time, and users must [explicitly enable them].
This allows accessing new functionality that is not yet stable without unwittingly relying on them.

[explicitly enable them]: ../nix-conf.md#conf-expeirmental-features

# When should a new feature be marked experimental?

A change in the Nix codebase should be guarded by an experimental feature flag if it is likely to either introduce a regression of any kind, or if it implies a non-trivial change to the external interface of Nix.

Examples:

- Changes to the Nix language, such as new built-ins, syntactic or semantic changes, etc.
- Changes to the command-line interface

# Lifecycle of an experimental feature

Experimental features have to be treated on a case-by-case basis.
However, the standard workflow for an experimental feature is as follows:

- The feature is implemented in a pull request.
- It is merged under an experimental flag, and is disabled by default.
    - Being experimental, the feature can still be changed arbitrarily or removed again.

> - After at least one release cycle without any significant issues, the feature becomes enabled by default.
>     - If any blocking issue gets reported the feature can be disabled by default again.
> - After at least another release cycle, the feature flag is removed and the feature is considered stable.

Reviewer: ```suggestion
- The experimental feature flag can be removed and the feature is considered stable if:
  - There is enough evidence of users having tried the feature, such as feedback, fixed bugs, demonstrations of how it is put to use
  - Maintainers feel confident that
    - The feature is designed and implemented sensibly, that it is fit for purpose
    - Potential interactions are well-understood
    - Stabilising the feature will not incur an outsized maintenance burden in the future
```

The following diagram illustrates the process:

<!-- TODO: replace with ASCII art to render correctly once contents are agreed upon -->

```mermaid
flowchart
    xp_disabled[experimental, disabled by default]
    xp_enabled[experimental, enabled by default]
    stable([stable])
    trunk([pull request])
    dropped([removed])

    trunk --> |merge into master| xp_disabled
    xp_disabled --> |At least one release without significant issues| xp_enabled
    xp_enabled --> |At least one release without significant issues| stable
    xp_enabled --> |Blocking issue reported| xp_disabled
    xp_disabled --> |"(breaking) changes"| xp_disabled

>     xp_disabled --> |Decision against adding the feature| dropped

Reviewer: Update the diagram to match the text description

This section describes the notion of *experimental features*, and how it fits into the big picture of the development of Nix.

# What are experimental features?

[Since Nix 2.4](../release-notes/rl-2.4.md), some features of Nix can be flagged as experimental.
Experimental features can be changed or removed at any time, and users must [explicitly enable them].
This allows accessing new functionality that is not yet stable without unwittingly relying on them.

> [explicitly enable them]: ../nix-conf.md#conf-expeirmental-features

Reviewer 2: ```suggestion
[explicitly enable them]: ../nix-conf.md#conf-experimental-features
```

Reviewer 2: BTW I'm really opposed to the "footnote" style of linking. It requires duplicating text like "explicitly enable them", meaning that if you change that text, you have to remember to update it in two places and not make a typo. It's much less verbose and error-prone to just write:

> ... and users must [explicitly enable them](../nix-conf.md#conf-experimental-features).

# When should a new feature be marked experimental?

> A change in the Nix codebase should be guarded by an experimental feature flag if it is likely to either introduce a regression of any kind, or if it implies a non-trivial change to the external interface of Nix.

Reviewer 2: I would drop the "introduce a regression of any kind". Any change can introduce a regression, but that's a matter of testing.

Reviewer 3: Is it the intent that non-trivial API changes will be rolled out first as experimental features?

Reviewer: ```suggestion
A change in the Nix codebase should be guarded by an experimental feature flag if it implies a backwards-incompatible or non-trivial change to the external interface of Nix.
```

Examples:

- Changes to the Nix language, such as new built-ins, syntactic or semantic changes, etc.
- Changes to the command-line interface

> - Some non-trivial internal changes that are likely to introduce bugs

Reviewer 2: ```suggestion
```

# Lifecycle of an experimental feature

Experimental features have to be treated on a case-by-case basis.
However, the standard workflow for an experimental feature is as follows:

- The feature is implemented in a pull request.
- It is merged under an experimental flag, and is disabled by default.
    - Being experimental, the feature can still be changed arbitrarily or removed again.

> - After at least one release cycle without any significant issues, the feature becomes enabled by default.

Reviewer 2: This does not reflect our current process. We don't enable experimental features ever, that's for the user to do.

Reviewer: That's true. Then we'd need to specify when to simply make a feature not experimental any more.

Just an idea:

> When maintainers are confident that a feature will not require breaking changes in the future, the experimental flag is removed and the feature considered stable.

Reviewer 4: I like the "cool down" period aspect, which is when we are ready to stabilize something we should make sure it hasn't changed since the last release. That means we might focus on something for stabilization, do some "finishing touches", and then wait a cycle.

Reviewer 5: For that stabilization cycle it is is more of a "feature flag" than an "experimental flag". This would make the feature opt-out vs opt-in.

Author: Yeah, calling it “experimental” is wrong here. I think it's a worthwile phase to have for some changes though, where we really want to flip the switch but can't be sure that we won't break something. But maybe we can save it for later

> This section aims at describing the notion of a Nix "experimental
> flag", and how it fits into the big picture of the developpement of Nix.

Reviewer 2: ```suggestion
This section describes the notion of a Nix's "experimental
features", and how it fits into the big picture of the development of Nix.
```

And rename "experimental/xp flags" to "experimental features" throughout, because the config setting is `experimental-features` after all.

Author: Yes, I renamed it to “experimental flag” because it doesn’t have to be a “feature” strictly speaking (because of the section on refactorings). But we can keep the “feature” word for consistency

(Or we could add a new `xp-feature-is-flag` experimental feature to rename the option to `experimental-flags`. But that’s going a bit too meta :wink: )

# What are these xp flags?

> Since Nix 2.4, some features of Nix can be flagged as
> experimental.Conceptually, this means that these features are not
> considered as part of Nix proper, but are only included in the code to
> make their developpement and testing easier.Practically, these require
> an explicit opt-in, by adding the name of the experimental feature to
> the `experimental-features` configuration option. This is really handy
> as it makes it harder to accidentally use an experimental feature
> without knowing it.

Reviewer 2: ```suggestion
Since Nix 2.4, some features of Nix can be flagged as
experimental. This means that they can be changed or removed at any time.
Users must explicitly opt in to them by adding the name of the experimental feature to
the `experimental-features` configuration option. This is really handy
as it makes it harder to accidentally use an experimental feature
without knowing it.
```

# When should an experimental flag be used

A change in the Nix codebase should be guarded by an experimental flag
if it is likely to either introduce a regression of any kind, or if it
changes the external interface of Nix (be it the language, the cli or

> anything else that's considered as "public").

Reviewer 2: I don't think everything that changes the external interface requires an experimental feature. The main criterion is whether we're uncertain about whether the feature is a good idea at all, or about the exact interface, or the implementation is not sufficiently stable yet. But we wouldn't make (for instance) a new builtin function experimental unless we have doubts about it.

Author: Fair point. What about:

```suggestion
A change in the Nix codebase should be guarded by an experimental flag
if it is likely to either introduce a regression of any kind, or if it
implies a non-trivial change in the external interface of Nix (be it the language, the cli or
anything else that's considered as "public").
```
?

(For the specific “builtin” case, I’d actually be in favor of making them experimental because that doesn’t cost much and can avoid having to commit to them forever if they prove not to be that useful. But that’s not central)

For example, these changes need to be guarded by an experimental flag:

-   Any language change (new builtin, syntactic or semantic change,
    etc..),

-   Any change to the (non-experimental) CLI interface

> -   A non-trivial internal refactoring that's likely to introduce some

Reviewer 2: By their nature, I don't think it's feasible to make refactorings experimental. Because that would mean you have to keep the old and new code around, which goes against the idea of a refactoring. Also, refactorings shouldn't break anything by definition.

Author: > shouldn’t

That’s where the devil lies :stuck_out_tongue: 

I guess it depends how you define “refactoring”. Something I had in mind in particular here is changes to the daemon protocol (say adding a new message type to make some specific operation faster), which _shouldn’t_ break anything, but happen to do so every once in a while (and for which we need to keep both versions anyways for backwards compatibility). Or just a re-implementation of a specific piece of functionality that’s well-contained (so cheap to duplicate for a short amount of time), but might be hiding some subtle bug

    bugs\[\^1\]

\[\^1\]: Having to be weighted against the costs of course. If the
refactoring changes a low-level interface and keeping both versions
around essentially means duplicating half of the codebase, then it might
not be worth doing that.

> # Life flow of an experimental flag

Reviewer 2: ```suggestion
# Lifecycle of an experimental feature
```

Experimental features being, well, experimental, they all have to be
treated on a case-by-case basis.However, the standard workflow for an
experimental flag is as follows:

-   The feature is merged, under an experimental flag (possibly over a
    long period of time)

-   It becomes part of a release

-   After at least a full 6w release cycle, the feature becomes enabled
    by default

-   After at lesat another full 6w release cycle, the old behavior is
    removed

This is obviously just indicative. In particular the third step is only
useful if the feature is susceptible to introduce some significant and
hard to work-around regression (for example the addition of a new

> language builtin). Conversely, the second is useless if people have no

Reviewer 2: This is unclear, because why would the addition of a new builtin cause a regression?

Author: Indeed. I think I first wrote the sentence the other way, so that was actually an exemple of something that’s unlikely to warrant this third step :upside_down_face:

> hard to work-around regression (for example the addition of a new
> language builtin). Conversely, the second is useless if people have no

Author: ```suggestion
hard to work-around regression (for example if it requires some non-trivial changes in existing codepaths). Conversely, the second is useless if people have no
```

compelling reason to switch the flag (for example a purely internal
refactoring)

> # Relation to the RFC process

Reviewer 2: I would leave out this section, since the relationship between RFCs and experimental features is itself something that should be clarified in a future RFC.

Author: I’ve some mixed feelings about that.

That admittedly falls in a grey area. This whole “experimental features” thing is technically an implementation detail, so something that should just be of the resort of the Nix development team. But at the same time there’s no such team (since the “core team” got disbanded).

But in that matter you’re the doc, doc. So if you feel like it, I’ll re-submit the whole thing as an RFC
